process/interceptors/peerHeartbeatInterceptor.go

```diff
@@ -0,0 +1,127 @@
+	peerHeartbeatProcessor           process.PeerHeartbeatProcessor
+}
+
+// NewPeerHeartbeatInterceptor hooks a new interceptor for packed multi data containing peer heartbeat instances
+func NewPeerHeartbeatInterceptor(arg ArgPeerHeartbeatInterceptor) (*peerHeartbeatInterceptor, error) {
+	err := checkArguments(arg.ArgSingleDataInterceptor)
+	if err != nil {
+		return nil, err
+	}
+	if check.IfNil(arg.Marshalizer) {
+		return nil, process.ErrNilMarshalizer
+	}
+	if check.IfNil(arg.ValidatorChecker) {
+		return nil, process.ErrNilValidatorChecker
+	}
+	if check.IfNil(arg.ProcessingThrottler) {
+		return nil, process.ErrNilProcessingThrottler
+	}
+	if check.IfNil(arg.HeartbeatProcessor) {
+		return nil, process.ErrNilHeartbeatProcessor
+	}
+
+	interceptor := &peerHeartbeatInterceptor{
+		baseDataInterceptor: &baseDataInterceptor{
+			throttler:        arg.Throttler,
+			antifloodHandler: arg.AntifloodHandler,
+			topic:            arg.Topic,
+			currentPeerId:    arg.CurrentPeerId,
+			processor:        arg.Processor,
+			debugHandler:     resolver.NewDisabledInterceptorResolver(),
+			marshalizer:      arg.Marshalizer,
+			factory:          arg.DataFactory,
+		},
+		validatorChecker:                 arg.ValidatorChecker,
+		peerHeartbeatProcessingThrottler: arg.ProcessingThrottler,
+		peerHeartbeatProcessor:           arg.HeartbeatProcessor,
+	}
+
+	return interceptor, nil
+}
+
+// ProcessReceivedMessage is the callback func from the p2p.Messenger and will be called each time a new message was received
+// (for the topic this validator was registered to)
+func (phi *peerHeartbeatInterceptor) ProcessReceivedMessage(message p2p.MessageP2P, fromConnectedPeer core.PeerID) error {
+	multiDataBuff, err := phi.preProcessMessage(message, fromConnectedPeer)
+	if err != nil {
+		return err
+	}
+
+	peerHbs := make([]process.InterceptedPeerHeartbeat, 0, len(multiDataBuff))
+	for _, dataBuff := range multiDataBuff {
+		var interceptedData process.InterceptedData
+		interceptedData, err = phi.interceptedData(dataBuff, message.Peer(), fromConnectedPeer)
+		if err != nil {
+			phi.throttler.EndProcessing()
+			return err
+		}
+
+		peerHb, ok := interceptedData.(process.InterceptedPeerHeartbeat)
+		if !ok {
+			//intercepted data is not of type interceptedPeerInfo
+			cause := "intercepted data is not of type process.InterceptedPeerInfo"
+			phi.blackListPeers(cause, nil, message.Peer(), fromConnectedPeer)
+			phi.throttler.EndProcessing()
+
+			return errors.New(cause)
+		}
+
+		var shardID uint32
+		_, shardID, err = phi.validatorChecker.GetValidatorWithPublicKey(peerHb.PublicKey())
+		isSkippableObserver := err != nil && !phi.peerHeartbeatProcessingThrottler.CanProcess()
+		if isSkippableObserver {
+			continue
+		}
+
+		//validators will always get the chance to get their message processed despite the number of observers messages
+		//being broadcast
+		phi.peerHeartbeatProcessingThrottler.StartProcessing()
+		peerHb.SetShardID(shardID)
```

Comment:
for an Observer will the shardID be set always to 0 if the CanProcess is true? The GetValidatorWithPublicKey returns nil, 0, error.

Reply:
Exactly. We should set the shard ID only for the validator nodes. Will change

Reply:
ok

Reply:
I think it would be important to set for obsevers as well as they route a lot of messages when they are connected. Could we get that information from the heartbeat ?

Reply:
Yes, actually the peer heartbeat structure will contain 2 shardID values (exactly in the same way the current implementation have 2 shardID fields): one for computed shardID and one for the advertised shardID (aka the shardID send from that peer through the peer heartbeat message). Will rename as to avoid confusions.

Reply:
done